Extended Pluggable Monitor specification to allow board-specific settings to be applied

cmaglie opened the pull request:
**Please check if the PR fulfills these requirements**

- [X] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [X] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)
- [ ] `UPGRADING.md` has been updated with a migration guide (for breaking changes)

* **What kind of change does this PR introduce?**
  Adds the possibility for board developers to override the default port settings when the monitor is opened.
  This is useful in particular for the `serial` port settings of `dtr` and `rts` since the Arduino IDE 1.8.x allowed the following `board.txt` properties:
  ```
  BOARD_ID.serial.disableRTS=true
  BOARD_ID.serial.disableDTR=true
  ```
  To generalize the above to the Pluggable Monitor a new way to override specific port settings has been added:
  ```
  BOARD_ID.monitor_port.PROTOCOL.SETTING_NAME=SETTING_VALUE
  ```
  where:
  * `BOARD_ID` is the board identifier
  * `PROTOCOL` is the port protocol
  * `SETTING_NAME` and `SETTING_VALUE` are the port setting and the desired value

  this is general enough to cover also the old use case (`dtr` and `rts`).
  For backward compatibility we will automatically convert the old directives `disableRTS` and `disableDTR` to the new one.

- **What is the current behavior?**
  The directives:
  ```
  BOARD_ID.serial.disableRTS=true
  BOARD_ID.serial.disableDTR=true
  ```
  are ignored.

* **What is the new behavior?**
  The directives above are followed.

- **Does this PR introduce a breaking change, and is [titled accordingly](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#breaking)?**
  No

* **Other information**:
  Related:
  * issues with serial monitor related to wrong `rts` and `dtr` settings: https://github.com/arduino/serial-monitor/issues/17
  * a PR that adds support for setting `rts` and `dtr` on the serial-monitor: https://github.com/arduino/serial-monitor/pull/21

per1234 commented on docs/platform-specification.md:
````diff
@@ -1191,6 +1191,47 @@ If the **upload.protocol** property is not defined for a board, the Arduino IDE'
+BOARD_ID.serial.disableDTR=true
+```
+
+to disable RTS and DTR when opening the serial monitor. The keep backward compatibilty the properties above are
````
```suggestion
to disable RTS and DTR when opening the serial monitor. To keep backward compatibility the properties above are
```
Typos

per1234 reviewed (changes requested):
### Describe the problem

Pluggable monitor port configuration properties are not recognized when set via [custom board options](https://arduino.github.io/arduino-cli/dev/platform-specification/#custom-board-options)

### To reproduce

Create a board definition that sets pluggable monitor port configuration properties via custom board options:

```text
menu.rts=RTS
menu.dtr=DTR
foo.name=Configurable
foo.menu.rts.on=On
foo.menu.rts.on.monitor_port.serial.rts=on
foo.menu.rts.off=Off
foo.menu.rts.off.monitor_port.serial.rts=off
foo.menu.dtr.on=On
foo.menu.dtr.on.monitor_port.serial.dtr=on
foo.menu.dtr.off=Off
foo.menu.dtr.off.monitor_port.serial.dtr=off
```

Set the option via the `--fqbn` argument of an `arduino-cli monitor` command.

🐛 The port is not configured by the property.

### Expected behavior

Pluggable monitor port configuration properties can be set via custom board options.

If this is not going to be supported, then the deficiency must be documented in the Arduino Platform Specification.

### Arduino CLI version

64999507

### Operating system

Windows 10

### Additional context

Arduino IDE 1.x recognizes the properties as expected when set via custom board options:

```text
menu.rts=RTS
menu.dtr=DTR
bar.menu.rts.on=On
bar.menu.rts.on.serial.disableRTS=false
bar.menu.rts.off=Off
bar.menu.rts.off.serial.disableRTS=true
bar.menu.dtr.on=On
bar.menu.dtr.on.serial.disableDTR=false
bar.menu.dtr.off=Off
bar.menu.dtr.off.serial.disableDTR=true
```

cmaglie commented on docs/platform-specification.md:
````diff
@@ -1191,6 +1191,47 @@ If the **upload.protocol** property is not defined for a board, the Arduino IDE'
+```
+
+The setting name and value can be anything available in the
+[pluggable monitor settings](pluggable-monitor-specification.md#describe-command).
````
I've rephrased it, I don't know if it's better... feel free to suggest any change :-)

per1234 replied:
Looks great. Thanks!

cmaglie commented:
> Pluggable monitor port configuration properties are not recognized when set via [custom board options](https://arduino.github.io/arduino-cli/dev/platform-specification/#custom-board-options)

I rebased the PR and fixed this issue, now the submenu options are supported.
This helped also to fix a weird edge case.

per1234 commented on docs/platform-specification.md:
````diff
@@ -885,6 +885,30 @@ may contain any extra parameter in the formula: this is different from the monit
 
 We strongly recommend using this syntax only for development purposes and not on released platforms.
 
+#### Port configuration
+
+Each pluggable monitor has its own default settings that can be overridden using the following board properties:
+
+```
+BOARD_ID.monitor_port.PROTOCOL.SETTING_NAME=SETTING_VALUE
+```
+
+where:
+
+- `BOARD_ID` is the board identifier
+- `PROTOCOL` is the port protocol
+- `SETTING_NAME` and `SETTING_VALUE` are the port setting and the desired value
+
+For example, let's suppose that a board needs the `baudrate` setting of the `serial` port to be `9600`, then the
+corresponding properties in the `boards.txt` file will be:
+
+```
+myboard.monitor_port.serial.baudrate=9600
+```
+
+The settings available in a specific pluggable monitor can be
+[queried directly from it](pluggable-monitor-specification.md#describe-command).
+
````
```suggestion
```
Move to the end of the section

per1234 commented on docs/platform-specification.md:
```diff
@@ -903,6 +927,23 @@ will allow all legacy non-pluggable platforms to migrate to pluggable monitor wi
 
 For detailed information, see the [Pluggable Monitor specification](pluggable-monitor-specification.md).
 
```
````suggestion

#### Port configuration

Each pluggable monitor has its own default settings that can be overridden using the following board properties:

```
BOARD_ID.monitor_port.PROTOCOL.SETTING_NAME=SETTING_VALUE
```

where:

- `BOARD_ID` is the board identifier
- `PROTOCOL` is the port protocol
- `SETTING_NAME` and `SETTING_VALUE` are the port setting and the desired value

For example, let's suppose that a board needs the `baudrate` setting of the `serial` port to be `9600`, then the
corresponding properties in the `boards.txt` file will be:

```
myboard.monitor_port.serial.baudrate=9600
```

The settings available in a specific pluggable monitor can be
[queried directly from it](pluggable-monitor-specification.md#describe-command).

````

Companion to the suggestion above.

per1234 reviewed (approved):
Thanks Cristian!
